<a href="https://colab.research.google.com/github/Shaunak-Mukherjee/ECE-59500---Reinforcement-Learning--Theory-and-Algorithms/blob/main/ECE_59500_Reinforcement_Learning_HW6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Solution to Problem 4.1**

In [5]:
import numpy as np
from math import exp, log
from IPython.display import Markdown

GAMMA = 0.95

# Rewards stored as dict (states/actions are text)

R = {
  'empty': {'stay': -1, 'explore': 1},
  'monster': {'stay': -2, 'evade': 3, 'befriend': -5},
  'food': {'stay': -2, 'explore': -2, 'collect': 5},
  'resource': {'stay': -2, 'explore': -2, 'collect': 4}
}


# Each trajectory is a list of (state, action) tuples
tau1 = [
  ('empty', 'stay'),
  ('monster', 'evade'),
  ('resource', 'collect')
]

tau2 = [
  ('empty', 'explore'),
  ('food', 'collect'),
  ('monster', 'befriend')
]

tau3 = [
  ('empty', 'explore'),
  ('empty', 'explore'),
  ('resource', 'collect')
]

trajectories = [tau1, tau2, tau3]

# Discounted returns
def calculate_discounted_return(trajectory, gamma, reward_map):
  """Calculates the discounted return G(tau) for a given trajectory."""
  G = 0.0
  for t, (state, action) in enumerate(trajectory):
    reward = reward_map[state][action]
    discount_factor = gamma ** t
    G += discount_factor * reward
  return G

G1 = calculate_discounted_return(tau1, GAMMA, R)
G2 = calculate_discounted_return(tau2, GAMMA, R)
G3 = calculate_discounted_return(tau3, GAMMA, R)
G = [G1, G2, G3]

# Output
out = f"""
4.1 Discounted Returns
- $\\tau_1$: {G1:.4f}
- $\\tau_2$: {G2:.4f}
- $\\tau_3$: {G3:.4f}
"""

Markdown(out)



4.1 Discounted Returns
- $\tau_1$: 5.4600
- $\tau_2$: 1.2375
- $\tau_3$: 5.5600


Solution to Problem 4.2

This is calculated using Bradley-Terry preference probability in Lecture 26.

In [6]:
import math
def pref_formula(i, j):
    """
    B-Terry formula:
    P(τ_i ≻ τ_j) = exp(G(τ_i)) / (exp(G(τ_i)) + exp(G(τ_j)))
    """
    Gi = G[i-1]
    Gj = G[j-1]

    return exp(Gi) / (exp(Gi) + exp(Gj))

P_1_over_2 = pref_formula(1, 2)

out = f"""
4.2 Probability human prefers $\\tau_1$ over $\\tau_2$

Using:

$
P(\\tau_1 \\succ \\tau_2)
=
\\frac{{\\exp(G(\\tau_1))}}{{\\exp(G(\\tau_1)) + \\exp(G(\\tau_2))}}
$

**Result:** {P_1_over_2:.4f}
"""

Markdown(out)


4.2 Probability human prefers $\tau_1$ over $\tau_2$

Using:

$
P(\tau_1 \succ \tau_2)
=
\frac{\exp(G(\tau_1))}{\exp(G(\tau_1)) + \exp(G(\tau_2))}
$

**Result:** 0.9855


Solution to Problem 4.3

In [7]:
# Human chooses tau1 over tau3 is:
P_1_over_3 = pref_formula(1, 3)

out = f"""
4.2 Probability human prefers $\\tau_1$ over $\\tau_3$

Using:

$
P(\\tau_1 \\succ \\tau_3)
=
\\frac{{\\exp(G(\\tau_1))}}{{\\exp(G(\\tau_1)) + \\exp(G(\\tau_3))}}
$

**Result:** {P_1_over_3:.4f}
"""

Markdown(out)


4.2 Probability human prefers $\tau_1$ over $\tau_3$

Using:

$
P(\tau_1 \succ \tau_3)
=
\frac{\exp(G(\tau_1))}{\exp(G(\tau_1)) + \exp(G(\tau_3))}
$

**Result:** 0.4750


Solution to Problem 4.4

In [8]:
# Compute the (maximum likelihood estimation) loss function
P_1_over_2 = pref_formula(1, 2) # Computed in 4.2
P_3_over_1 = pref_formula(3, 1) # Probability human prefers $\\tau_3$ over $\\tau_1$
loss = - (log(P_1_over_2) + log(P_3_over_1))

out = f"""
MLE Computation Results
Preference probabilities:
- P(τ₁ ≻ τ₂) = {P_1_over_2:.4f}
- P(τ₃ ≻ τ₁) = {P_3_over_1:.4f}

MLE Loss:
- **Result** Loss = {loss:.4f}
"""

Markdown(out)


MLE Computation Results
Preference probabilities:
- P(τ₁ ≻ τ₂) = 0.9855
- P(τ₃ ≻ τ₁) = 0.5250

MLE Loss:
- **Result** Loss = 0.6590


In [9]:
P_2_over_1 = pref_formula(2, 1)
print(P_2_over_1)

0.014450077563669653


**Solution to Problem 4.5**

If the human picks $\tau_2 \succ \tau_1$ and $\tau_3 \succ \tau_1$:The Negative Log-Likelihood loss function is expected to increase significantly.
Because $G(\tau_1) \gg G(\tau_2)$ as we have seen in earlier computation 4.1, the Bradley–Terry model computes:

$
P(\tau_1 \succ \tau_2) \approx 0.986,\quad
P(\tau_2 \succ \tau_1) \approx 0.014.
$

Also as computed earlier the human’s label $\tau_1$ ≻ $\tau_2$ matched the model’s high-probability prediction, resulting in a small loss term $
-\log P(\tau_1 \succ \tau_2)
$
= 0.65
If the human instead reports $\tau_2$ ≻ $\tau_1$, the loss term becomes $
-\log P(\tau_2 \succ \tau_1),$ which is much larger because the probability extremely low.
Since the $\tau_3$ ≻ $\tau_1$ label remains unchanged, the total loss strictly increases.
